collector/infiniband_linux.go

```diff
@@ -0,0 +1,167 @@
+	"github.com/prometheus/common/log"
+)
+
+const infinibandPath = "class/infiniband"
+
+var (
+	infinibandNoDevicesFound = errors.New("InfiniBand / No devices detected")
+	infinibandNoPortsFound   = errors.New("InfiniBand / No ports detected")
+)
+
+type infinibandCollector struct {
+	metricDescs map[string]*prometheus.Desc
+	counters    map[string]string
+}
+
+func init() {
+	Factories["infiniband"] = NewInfiniBandCollector
+}
+
+func NewInfiniBandCollector() (Collector, error) {
+	var i infinibandCollector
+
+	// Filenames of all InfiniBand counter metrics including a detailed description.
+	i.counters = map[string]string{
+		"link_downed":            "Number of times the link failed to recover from an error state and went down",
+		"link_error_recovery":    "Number of times the link successfully recovered from an error state",
+		"multicast_rcv_packets":  "Number of multicast packets received (including errors)",
+		"multicast_xmit_packets": "Number of multicast packets transmitted (including errors)",
+		"port_rcv_data":          "Number of data octets received on all links",
+		"port_xmit_data":         "Number of data octets transmitted on all links",
+		"unicast_rcv_packets":    "Number of unicast packets received (including errors)",
+		"unicast_xmit_packets":   "Number of unicast packets transmitted (including errors)",
+	}
+
+	subsystem := "infiniband"
+	i.metricDescs = make(map[string]*prometheus.Desc)
+
+	for metric, description := range i.counters {
+		i.metricDescs[metric] = prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, metric+"_total"),
+			description,
+			[]string{"device", "port"},
+			nil)
+	}
+
+	return &i, nil
+}
+
+// Retrieve a list of InfiniBand devices.
+func infinibandDevices(infinibandPath string) ([]string, error) {
+	devices, err := filepath.Glob(filepath.Join(infinibandPath, "/*"))
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(devices) < 1 {
+		log.Debugf("Unable to detect InfiniBand devices")
+		err = infinibandNoDevicesFound
+		return nil, err
+	}
+
+	// Extract just the filenames which equate to the device names.
+	for i, device := range devices {
+		devices[i] = filepath.Base(device)
+	}
+
+	return devices, nil
+}
+
+// Retrieve a list of ports for the InfiniBand device.
+func infinibandPorts(infinibandPath, device string) ([]string, error) {
+	ports, err := filepath.Glob(filepath.Join(infinibandPath, device, "ports/*"))
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ports) < 1 {
+		log.Debugf("Unable to detect ports for %s", device)
+		err = infinibandNoPortsFound
+		return nil, err
+	}
+
+	// Extract just the filenames which equates to the port numbers.
+	for i, port := range ports {
+		ports[i] = filepath.Base(port)
+	}
+
+	return ports, nil
+}
+
+func readMetric(directory, metricFile string) (uint64, error) {
+	metric, err := readUintFromFile(filepath.Join(directory, metricFile))
+
+	if err != nil {
+		log.Debugf("Error reading " + metricFile + " file")
```

Comment:
Since these are "formatted printing" log lines, IMO it's a little cleaner to not use concatentation: `log.Debugf("Error reading file %q", metricFile)`.

`%q` is nice because it nicely quotes the string for you.